File: frontend/src/lib/queryExpansion.ts
const QUERY_EXPANSIONS: Record<string, string[]> = {
  luxury: ["designer", "premium", "high quality"],
  premium: ["luxury", "designer", "high quality"],
  designer: ["luxury", "premium"],
  handbag: ["bag", "purse", "leather handbag", "designer handbag"],
  handbags: ["bag", "purse", "leather handbag", "designer handbag"],
  bag: ["handbag", "purse"],
  watch: ["timepiece", "wrist watch", "premium watch"],
  watches: ["timepiece", "wrist watch", "premium watch"],
  formal: ["office", "business", "dressy"],
  office: ["formal", "business"],
  shirt: ["shirts", "topwear", "button down"],
  shirts: ["shirt", "topwear", "button down"],
  tshirt: ["t-shirt", "tee", "casual top"],
  tshirts: ["t-shirt", "tee", "casual top"],
  "t-shirt": ["tshirt", "tee", "casual top"],
  jeans: ["denim", "blue jeans"],
  shoes: ["footwear"],
  sunglasses: ["shades", "eyewear"],
  kurta: ["ethnic wear", "traditional"],
  kurtas: ["ethnic wear", "traditional"],
  dress: ["dresses", "gown", "party wear"],
  dresses: ["dress", "gown", "party wear"],
};

export function tokenize(text: string): string[] {
  return text.toLowerCase().match(/[a-z0-9]+/g) ?? [];
}

export function expandQuery(query: string, maxAddedTerms = 12): string {
  const normalizedQuery = query.trim().replace(/\s+/g, " ");
  if (!normalizedQuery) return "";

  const additions: string[] = [];
  const seen = new Set<string>([
    normalizedQuery.toLowerCase(),
    ...tokenize(normalizedQuery),
  ]);

  for (const token of tokenize(normalizedQuery)) {
    for (const expansion of QUERY_EXPANSIONS[token] ?? []) {
      const normalizedExpansion = expansion.toLowerCase();
      if (seen.has(normalizedExpansion)) continue;

      additions.push(expansion);
      seen.add(normalizedExpansion);
      if (additions.length >= maxAddedTerms) {
        return [normalizedQuery, ...additions].join(" ");
      }
    }
  }

  return [normalizedQuery, ...additions].join(" ");
}

File: backend/query_expansion.py
"""
Lightweight query expansion and tokenization for retrieval.

The expansion dictionary is intentionally small and deterministic. It improves
recall for common fashion search intents without introducing LLM latency or
unbounded query drift.
"""

from __future__ import annotations

import re


QUERY_EXPANSIONS = {
    "luxury": ["designer", "premium", "high quality"],
    "premium": ["luxury", "designer", "high quality"],
    "designer": ["luxury", "premium"],
    "handbag": ["bag", "purse", "leather handbag", "designer handbag"],
    "handbags": ["bag", "purse", "leather handbag", "designer handbag"],
    "bag": ["handbag", "purse"],
    "watch": ["timepiece", "wrist watch", "premium watch"],
    "watches": ["timepiece", "wrist watch", "premium watch"],
    "formal": ["office", "business", "dressy"],
    "office": ["formal", "business"],
    "shirt": ["shirts", "topwear", "button down"],
    "shirts": ["shirt", "topwear", "button down"],
    "tshirt": ["t-shirt", "tee", "casual top"],
    "tshirts": ["t-shirt", "tee", "casual top"],
    "t-shirt": ["tshirt", "tee", "casual top"],
    "jeans": ["denim", "blue jeans"],
    "shoes": ["footwear"],
    "sunglasses": ["shades", "eyewear"],
    "kurta": ["ethnic wear", "traditional"],
    "kurtas": ["ethnic wear", "traditional"],
    "dress": ["dresses", "gown", "party wear"],
    "dresses": ["dress", "gown", "party wear"],
}


def tokenize(text: str) -> list[str]:
    """Tokenize text for BM25 with predictable punctuation handling."""
    return re.findall(r"[a-z0-9]+", text.lower())


def expand_query(query: str, max_added_terms: int = 12) -> str:
    """Return query plus bounded fashion synonyms for embedding and BM25."""
    query = " ".join(str(query).split())
    if not query:
        return ""

    tokens = tokenize(query)
    additions: list[str] = []
    seen = {query.lower(), *tokens}

    for token in tokens:
        for expansion in QUERY_EXPANSIONS.get(token, []):
            normalized = expansion.lower()
            if normalized in seen:
                continue
            additions.append(expansion)
            seen.add(normalized)
            if len(additions) >= max_added_terms:
                return " ".join([query, *additions])

    return " ".join([query, *additions])
